src/xunit.performance.api/SafeTerminateHandler.cs

```diff
@@ -0,0 +1,104 @@
+﻿using Microsoft.Xunit.Performance.Api.Native.Windows;
+using System;
+
+namespace Microsoft.Xunit.Performance.Api
+{
+    /// <summary>
+    /// Provides a mechanism for releasing disposable resources when CTRL+C,
+    /// CTRL+BREAK, or Close event (when the user closes the console by either
+    /// clicking Close on the console windows's window menu, or by clicking the
+    /// "End Task" button on the Task Manager) signals are received by the
+    /// application.
+    /// </summary>
+    /// <remarks>
+    /// 1. This class it is currently adding support for Windows only.
+    /// 2. This class owns the Disposable object that is being wrapped, and it is responsible for its cleanup.
+    /// </remarks>
+    /// <typeparam name="T">The type that this class encapsulates.</typeparam>
+    internal sealed class SafeTerminateHandler<T> : IDisposable
+        where T : class, IDisposable
+    {
+        /// <summary>
+        /// Initializes a new instance of the SafeTerminateHandler<T> class that
+        /// wraps a disposable object that will be created by calling the
+        /// specified callback Func.
+        /// </summary>
+        /// <param name="createCallback">A method that has no parameters and returns a new disposable object of the type specified by the T parameter.</param>
+        public SafeTerminateHandler(Func<T> createCallback)
+        {
+            _disposedValue = true;
+            _lock = new object();
+            BaseDisposableObject = null;
+            _HandlerRoutine = (Kernel32.CtrlTypes dwCtrlType) =>
+            {
+                switch (dwCtrlType)
+                {
+                    case Kernel32.CtrlTypes.CTRL_C_EVENT:
+                    case Kernel32.CtrlTypes.CTRL_BREAK_EVENT:
+                    case Kernel32.CtrlTypes.CTRL_CLOSE_EVENT:
+                        Dispose();
+                        break;
+                    default:
+                        break;
+                }
+                return false;
+            };
+
+            if (!Kernel32.SetConsoleCtrlHandler(_HandlerRoutine, true))
```

Comment:
Is this safe outside of the critical section? It's possible for the console ctrl handler to get created, but ctrl-C to be invoked before we enter the critical section below to set _disposedValue to false, and we never remove the handler. Even if this is a "safe" leak (the handler goes away when the process exits) I think it would better form to move this into the critical section.